<a href="https://colab.research.google.com/github/gptix/is-this-poisonous/blob/master/is_this_poisonous.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Can I Eat This Project
Lambda School Data Science cohort 10

## Requirements

See checklist at bottom of this notebook.

### At present, this notebook and other work are designed to fulfill the MVP requirements.

## Data
Data gotten from UCI **UC Irvine Machine Learning Repository**

https://archive.ics.uci.edu/ml/datasets/Mushroom


# Model Interpretation

- Visualize and interpret **partial dependence plots**
- Explain individual predictions with **shapley value plots**

### Three types of model explanations this unit:

#### 1. Global model explanation: all features in relation to each other _(Last lesson)_
- Feature Importances: _Default, fastest, good for first estimates_
- Drop-Column Importances: _The best in theory, but much too slow in practice_
- Permutaton Importances: _A good compromise!_

#### 2. Global model explanation: individual feature(s) in relation to target _(This lesson)_
- Partial Dependence plots

#### 3. Individual prediction explanation _(This lesson)_
- Shapley Values

_Note that the coefficients from a linear model give you all three types of explanations!_

### Links

#### Partial Dependence Plots
- [Kaggle / Dan Becker: Machine Learning Explainability — Partial Dependence Plots](https://www.kaggle.com/dansbecker/partial-plots)
- [Christoph Molnar: Interpretable Machine Learning — Partial Dependence Plots](https://christophm.github.io/interpretable-ml-book/pdp.html) + [animated explanation](https://twitter.com/ChristophMolnar/status/1066398522608635904)
- [pdpbox repo](https://github.com/SauceCat/PDPbox) & [docs](https://pdpbox.readthedocs.io/en/latest/)

#### Shapley Values
- [Kaggle / Dan Becker: Machine Learning Explainability — SHAP Values](https://www.kaggle.com/learn/machine-learning-explainability)
- [Christoph Molnar: Interpretable Machine Learning — Shapley Values](https://christophm.github.io/interpretable-ml-book/shapley.html)
- [SHAP repo](https://github.com/slundberg/shap) & [docs](https://shap.readthedocs.io/en/latest/)

### Setup

Run the code cell below. You can work locally (follow the [local setup instructions](https://lambdaschool.github.io/ds/unit2/local/)) or on Colab.

Libraries:

- category_encoders
- matplotlib
- numpy
- pandas
- [**pdpbox**](https://github.com/SauceCat/PDPbox)
- plotly
- scikit-learn
- scipy.stats
- [**shap**](https://github.com/slundberg/shap)
- xgboost


In [1]:
%%capture
import sys

# If on Colab:
if 'google.colab' in sys.modules:
    DATA_PATH = 'https://raw.githubusercontent.com/LambdaSchool/DS-Unit-2-Applied-Modeling/master/data/'
    !pip install category_encoders==2.*
    !pip install pdpbox
    !pip install shap

# If local:
else:
    DATA_PATH = '../data/'
    !pip install xgboost

# Ignore this warning: https://github.com/dmlc/xgboost/issues/4300
# xgboost/core.py:587: FutureWarning: Series.base is deprecated and will be removed in a future version
import warnings
warnings.filterwarnings(action='ignore', category=FutureWarning, module='xgboost')

## Ingest data and make it useful

In [2]:
import numpy as np
import pandas as pd

# I made the column names by copying descriptions from the file './data/agaricus-lepiota.names',
# then used lisp to extract column names, replace '-' with '_', then build a python list of strings.

colnames = ["poisonous_edible", "cap_shape", "cap_surface", "cap_color", "bruises?", "odor", "gill_attachment",
           "gill_spacing", "gill_size", "gill_color", "stalk_shape", "stalk_root", 
           "stalk_surface_above_ring", "stalk_surface_below_ring", "stalk_color_above_ring", 
           "stalk_color_below_ring", "veil_type", "veil_color", "ring_number", "ring_type", 
           "spore_print_color", "population", "habitat"]

# colnames

In [3]:
# a file named 'expanded' is a CSV file with 'full-word' values instead of single-letter abbreviations.
df = pd.read_csv(DATA_PATH+'expanded', skiprows=9, skipfooter=1, 
                 engine='python', names=colnames, header=None)
df.head()
# df.tail()
# df.isnull().sum()

poisonous_edible cap_shape cap_surface cap_color bruises?    odor  \
0           EDIBLE    CONVEX      SMOOTH     WHITE  BRUISES  ALMOND   
1           EDIBLE    CONVEX      SMOOTH     WHITE  BRUISES  ALMOND   
2           EDIBLE    CONVEX      SMOOTH     WHITE  BRUISES  ALMOND   
3           EDIBLE    CONVEX      SMOOTH     WHITE  BRUISES  ALMOND   
4           EDIBLE    CONVEX      SMOOTH     WHITE  BRUISES  ALMOND   

  gill_attachment gill_spacing gill_size gill_color  ...  \
0            FREE      CROWDED    NARROW      WHITE  ...   
1            FREE      CROWDED    NARROW      WHITE  ...   
2            FREE      CROWDED    NARROW       PINK  ...   
3            FREE      CROWDED    NARROW       PINK  ...   
4            FREE      CROWDED    NARROW      BROWN  ...   

  stalk_surface_below_ring stalk_color_above_ring stalk_color_below_ring  \
0                   SMOOTH                  WHITE                  WHITE   
1                   SMOOTH                  WHITE                  WHITE   
2                   SMOOTH                  WHITE                  WHITE   
3                   SMOOTH                  WHITE                  WHITE   
4                   SMOOTH                  WHITE                  WHITE   

  veil_type veil_color ring_number ring_type spore_print_color population  \
0   PARTIAL      WHITE         ONE   PENDANT            PURPLE    SEVERAL   
1   PARTIAL      WHITE         ONE   PENDANT             BROWN    SEVERAL   
2   PARTIAL      WHITE         ONE   PENDANT            PURPLE    SEVERAL   
3   PARTIAL      WHITE         ONE   PENDANT             BROWN    SEVERAL   
4   PARTIAL      WHITE         ONE   PENDANT            PURPLE    SEVERAL   

  habitat  
0   WOODS  
1   WOODS  
2   WOODS  
3   WOODS  
4   WOODS  

[5 rows x 23 columns]

## Split dataframe into **X** matrix and $\hat{y}$ vector

We will use the 22 features to 'predict' whether an observed mushroom is poisonous or not.

$\hat{y}$ is simply the series of 'POISONOUS' and 'EDIBLE' values.

And **X** is the matrix of all other series.

In [4]:
target = "poisonous_edible"
features = df.columns.drop(target)

X = df[features]
y = df[target]
# y

## Convert categorical values in $\hat{y}$ into numeric values.

In [5]:
# Transform string values to category values, then to integers
from sklearn import preprocessing
le = preprocessing.LabelEncoder()

le.fit(y)
list(le.classes_)

y = le.transform(y)

# Note: EDIBLE = 0, POISONOUS = 1
# since this mutates the data in y, transforming again will lose the relation between the old, string 
# data and the new, numeric data.
# To get back, re-run code at and after definition of df.

# Get sample of inversion of transformation.
list(le.inverse_transform([0, 1, 1]))

# df.head()

['EDIBLE', 'POISONOUS', 'POISONOUS']

## Split **X** and $\hat{y}$ into subsets for training, validation, and final testing.

In [6]:
from sklearn.model_selection import train_test_split
# Split train into train & val
# train, test = train_test_split(df, train_size=0.90, test_size=0.10, random_state=43)
# train, val = train_test_split(train, train_size=0.90, test_size=0.10, random_state=43)

# The 'train' data from this step will be split again, below.
X_train, X_test, y_train, y_test = train_test_split( X, y, test_size=0.10, random_state=42)

# In the line below, 'test_size' is used to specify the size of the val set.
X_train, X_val, y_train, y_val = train_test_split( X_train, y_train, test_size=0.11, random_state=42)

In [7]:
# print(len(X_train), len(X_val), len(X_test))
# pdy =pd.DataFrame(y)
# pdy

## Basic histogram


In [8]:
df['poisonous_edible'].hist();

## Get a baseline, to use in evaluating subsequent models.

In [9]:
# Get a baseline
baseline_edible = df['poisonous_edible'].value_counts(normalize=True).max()
print(f'Baseline, guessing \'edible\', would be right {round(100 * baseline_edible,1)}% of the time.\n')

Baseline, guessing 'edible', would be right 53.3% of the time.



## (Build and) Fit a Linear Regression model

In [10]:
# We need to convert categorical values to numeric values. 
import category_encoders as ce

# We will use a straightforward linear regression.
from sklearn.linear_model import LinearRegression

# We will use 'make_pipeline' to create a 'composition' of two classes.
from sklearn.pipeline import make_pipeline

# This instantiates and composes.  Really cool.
lr = make_pipeline(
    ce.TargetEncoder(),  
    LinearRegression()
)

# Use the instance of the model to fit to training features, and evaluate with training target output.
lr.fit(X_train, y_train)

Pipeline(memory=None,
         steps=[('targetencoder',
                 TargetEncoder(cols=['cap_shape', 'cap_surface', 'cap_color',
                                     'bruises?', 'odor', 'gill_attachment',
                                     'gill_spacing', 'gill_size', 'gill_color',
                                     'stalk_shape', 'stalk_root',
                                     'stalk_surface_above_ring',
                                     'stalk_surface_below_ring',
                                     'stalk_color_above_ring',
                                     'stalk_color_below_ring', 'veil_type',
                                     'veil_color', 'ring_number', 'ring_type',
                                     'spore_print_color', 'population',
                                     'habitat'],
                               drop_invariant=False, handle_missing='value',
                               handle_unknown='value', min_samples_leaf=1,
                       

# Evaluate Linear Model using **R^2**

In [11]:
print('Linear Regression R^2', lr.score(X_val, y_val))#

Linear Regression R^2 0.9644133999445964


### Explaining Linear Regression

In [12]:
coefficients = lr.named_steps['linearregression'].coef_
pd.Series(coefficients, features)

cap_shape                  -6.761478e-02
cap_surface                -4.069967e-03
cap_color                   6.904132e-02
bruises?                   -8.356050e-02
odor                        8.659845e-01
gill_attachment            -8.241986e-01
gill_spacing               -4.625803e-04
gill_size                   1.875690e-01
gill_color                  1.347181e-02
stalk_shape                -6.142119e-02
stalk_root                 -2.523471e-01
stalk_surface_above_ring    2.987629e-02
stalk_surface_below_ring   -3.658720e-02
stalk_color_above_ring      4.983177e-03
stalk_color_below_ring      3.828047e-02
veil_type                  -8.326673e-17
veil_color                  3.418248e-01
ring_number                -1.960591e-01
ring_type                  -9.430129e-03
spore_print_color           2.312993e-01
population                  1.107353e-01
habitat                    -4.841940e-02
dtype: float64

## (Build and) Fit a Tree Classifier

In [13]:
from sklearn.tree import DecisionTreeClassifier

pipeline = make_pipeline(
    ce.OrdinalEncoder(),
    DecisionTreeClassifier(max_depth=3)
)

pipeline.fit(X_train, y_train)

Validation Accuracy:  0.986810551558753


## Evaluate Accuracy of Tree Classifier

In [16]:
print('Validation Accuracy: ', pipeline.score(X_val, y_val))

Validation Accuracy:  0.986810551558753


## Calculate Recall Score of DecisionTreeClassifier

In [22]:
from sklearn.metrics import recall_score
y_pred = pipeline.predict(X_val)

# y_pred
recall_score = recall_score(y_val, y_pred)
print('Validation Recall: ', recall_score)

Validation Recall:  0.9842931937172775


In [2]:
!pip install graphviz

# Vizualize Decision Tree
import graphviz
from sklearn.tree import export_graphviz

tree = pipeline.named_steps['decisiontreeclassifier']

dot_data = export_graphviz(
    tree,
    out_file = None,
    feature_names = X_train.columns,
    class_names = y.unique().astype(str),
    filled = True,
    impurity=False,
    proportion=True,
)

graphviz.Source(dot_data)

ModuleNotFoundError: No module named 'graphviz'

### Fit Gradient Boosting model


In [14]:
!pip install xgboost
from sklearn.metrics import r2_score
from xgboost import XGBRegressor

gb = make_pipeline(
    ce.OrdinalEncoder(), 
    XGBRegressor(n_estimators=200, objective='reg:squarederror', n_jobs=-1)
)

gb.fit(X_train, y_train)
y_pred = gb.predict(X_val)
print('Gradient Boosting R^2', r2_score(y_val, y_pred))

ModuleNotFoundError: No module named 'xgboost'

### Explaining Gradient Boosting???

Linear models have coefficients, but tree ensembles do not.

Instaed, to see the relationship between individual feature(s) and the target, we can use partial dependence plots.

## Partial Dependence Plots


From [PDPbox documentation](https://pdpbox.readthedocs.io/en/latest/):

>**The common headache**: When using black box machine learning algorithms like random forest and boosting, it is hard to understand the relations between predictors and model outcome. For example, in terms of random forest, all we get is the feature importance. Although we can know which feature is significantly influencing the outcome based on the importance calculation, it really sucks that we don’t know in which direction it is influencing. And in most of the real cases, the effect is non-monotonic. We need some powerful tools to help understanding the complex relations between predictors and model prediction.


[Animation by Christoph Molnar](https://twitter.com/ChristophMolnar/status/1066398522608635904), author of [_Interpretable Machine Learning_](https://christophm.github.io/interpretable-ml-book/pdp.html#examples)

> Partial dependence plots show how a feature affects predictions of a Machine Learning model on average.
> 1. Define grid along feature
> 2. Model predictions at grid points
> 3. Line per data instance -> ICE (Individual Conditional Expectation) curve
> 4. Average curves to get a PDP (Partial Dependence Plot)


## Partial Dependence Plots with 1 feature

#### PDPbox
- [Gallery](https://github.com/SauceCat/PDPbox#gallery)
- [API Reference: pdp_isolate](https://pdpbox.readthedocs.io/en/latest/pdp_isolate.html)
- [API Reference: pdp_plot](https://pdpbox.readthedocs.io/en/latest/pdp_plot.html)


In [ ]:
# Later, when you save matplotlib images to include in blog posts or web apps,
# increase the dots per inch (double it), so the text isn't so fuzzy
import matplotlib.pyplot as plt
plt.rcParams['figure.dpi'] = 72

In [ ]:
from pdpbox.pdp import pdp_isolate, pdp_plot

feature = 'Annual Income'

isolated = pdp_isolate(
    model=gb, 
    dataset=X_val, 
    model_features=X_val.columns, 
    feature=feature
)

In [ ]:
gb

b

In [ ]:
pdp_plot(isolated, feature_name=feature);

In [ ]:
pdp_plot(isolated, feature_name=feature)
plt.xlim((20000,150000));

In [ ]:
# Plot PDP with 100 ICE curves
# PDP: Partial Dependence Plot
# ICE: Individual Conditional Expectation
pdp_plot(isolated, feature_name=feature, plot_lines=True, 
         frac_to_plot=0.01)

plt.xlim(20000,150000);

In [ ]:
isolated = pdp_isolate(
    model=gb, 
    dataset=X_val, 
    model_features=X_val.columns, 
    feature=feature, 
    num_grid_points=50
)

# How many predictions are required to make a PDP with 1 feature,
# given the size of our dataset, and the number of grid points?
len(X_val) * 50

In [ ]:
pdp_plot(isolated, feature_name=feature, plot_lines=True, 
         frac_to_plot=0.01)

plt.xlim(20000,150000);

## Partial Dependence Plots with 2 features

See interactions!

PDPbox
- [Gallery](https://github.com/SauceCat/PDPbox#gallery)
- [API Reference: pdp_interact](https://pdpbox.readthedocs.io/en/latest/pdp_interact.html)
- [API Reference: pdp_interact_plot](https://pdpbox.readthedocs.io/en/latest/pdp_interact_plot.html)

Be aware of a bug in PDPBox version <= 0.20:
- With the `pdp_interact_plot` function, `plot_type='contour'` gets an error, but `plot_type='grid'` works
- This will be fixed in the next release of PDPbox: https://github.com/SauceCat/PDPbox/issues/40


In [ ]:
from pdpbox.pdp import pdp_interact, pdp_interact_plot

In [ ]:
features = ['Annual Income', 'Credit Score']

interaction = pdp_interact(
    model=gb, 
    dataset=X_val,
    model_features=X_val.columns, 
    features=features
)

In [ ]:
pdp_interact_plot(interaction, plot_type='grid', 
                  feature_names=features);

### BONUS: 3D with Plotly!


In [ ]:
# First, make the 2D plot above. Then ...

pdp = interaction.pdp.pivot_table(
    values='preds', 
    columns=features[0], 
    index=features[1]
)[::-1] # Slice notation to reverse index order so y axis is ascending

In [ ]:
pdp = pdp.drop(columns=[1000.0, 751329.0])

pdp

In [ ]:
import plotly.graph_objs as go

surface = go.Surface(
    x=pdp.columns, 
    y=pdp.index, 
    z=pdp.values
)


layout = go.Layout(
    scene=dict(
        xaxis=dict(title=features[0]), 
        yaxis=dict(title=features[1]), 
        zaxis=dict(title=target)
    )
)

fig = go.Figure(surface, layout)
fig.show()

In [ ]:
# Number of predictions required to make a PDP with 2 feature
# given the size of your dataset, and the number of grid points:
len(X_val) * 10**2

# BONUS: PDPs with categorical features

1. I recommend you use Ordinal Encoder or Target Encoder, outside of a pipeline, to encode your data first. (If there is a natural ordering, then take the time to encode it that way, instead of random integers.) Then use the encoded data with pdpbox.
2. There's some extra work to get readable category names on your plot, instead of integer category codes.


In [ ]:
# Fit a model on Titanic data
import category_encoders as ce
import seaborn as sns
from sklearn.ensemble import RandomForestClassifier

df = sns.load_dataset('titanic')
df.age = df.age.fillna(df.age.median())
df = df.drop(columns='deck')
df = df.dropna()

target = 'survived'
features = df.columns.drop(['survived', 'alive'])

X = df[features]
y = df[target]

# Use Ordinal Encoder, outside of a pipeline
encoder = ce.OrdinalEncoder()
X_encoded = encoder.fit_transform(X)

model = RandomForestClassifier(n_estimators=100, random_state=42, n_jobs=-1)
model.fit(X_encoded, y)

In [ ]:
# Use Pdpbox
%matplotlib inline
import matplotlib.pyplot as plt
from pdpbox import pdp
feature = 'sex'
pdp_dist = pdp.pdp_isolate(model=model, dataset=X_encoded, model_features=features, feature=feature)
pdp.pdp_plot(pdp_dist, feature);

In [ ]:
# Look at the encoder's mappings
encoder.mapping

In [ ]:
pdp.pdp_plot(pdp_dist, feature)

# Manually change the xticks labels
plt.xticks([1, 2], ['male', 'female']);

In [ ]:
# Let's automate it

feature = 'sex'
for item in encoder.mapping:
    if item['col'] == feature:
        feature_mapping = item['mapping']
        
feature_mapping = feature_mapping[feature_mapping.index.dropna()]
category_names = feature_mapping.index.tolist()
category_codes = feature_mapping.values.tolist()

In [ ]:
pdp.pdp_plot(pdp_dist, feature)

# Automatically change the xticks labels
plt.xticks(category_codes, category_names);

In [ ]:
features = ['sex', 'age']

interaction = pdp_interact(
    model=model, 
    dataset=X_encoded, 
    model_features=X_encoded.columns, 
    features=features
)

pdp_interact_plot(interaction, plot_type='grid', feature_names=features);

In [ ]:
pdp = interaction.pdp.pivot_table(
    values='preds', 
    columns=features[0], # First feature on x axis
    index=features[1]    # Next feature on y axis
)[::-1]  # Reverse the index order so y axis is ascending

pdp = pdp.rename(columns=dict(zip(category_codes, category_names)))
plt.figure(figsize=(10,8))
sns.heatmap(pdp, annot=True, fmt='.2f', cmap='viridis')
plt.title('Partial Dependence of Titanic survival, on sex & age');

# Explain individual predictions with shapley value plots

## Regression: NYC Apartment Rents

_**Coming full circle!**_


In [ ]:
import numpy as np
import pandas as pd

# Read New York City apartment rental listing data
df = pd.read_csv(DATA_PATH+'apartments/renthop-nyc.csv')
assert df.shape == (49352, 34)

# Remove the most extreme 1% prices,
# the most extreme .1% latitudes, &
# the most extreme .1% longitudes
df = df[(df['price'] >= np.percentile(df['price'], 0.5)) & 
        (df['price'] <= np.percentile(df['price'], 99.5)) & 
        (df['latitude'] >= np.percentile(df['latitude'], 0.05)) & 
        (df['latitude'] < np.percentile(df['latitude'], 99.95)) &
        (df['longitude'] >= np.percentile(df['longitude'], 0.05)) & 
        (df['longitude'] <= np.percentile(df['longitude'], 99.95))]

# Do train/test split
# Use data from April & May 2016 to train
# Use data from June 2016 to test
df['created'] = pd.to_datetime(df['created'], infer_datetime_format=True)
cutoff = pd.to_datetime('2016-06-01')
train = df[df.created < cutoff]
test  = df[df.created >= cutoff]

_**Remember this code you wrote for your first assignment?**_

In [ ]:
# Arrange X features matrix & y target vector
features = ['bedrooms', 'bathrooms']
target = 'price'
X = df[features]
y = df[target]

# Fit model
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(X, y)

def predict(bedrooms, bathrooms):
    y_pred = model.predict([[bedrooms, bathrooms]])
    estimate = y_pred[0]
    bed_coef = model.coef_[0]
    bath_coef = model.coef_[1]
    
    # Format with $ and comma separators. No decimals.
    result = f'Rent for a {bedrooms}-bed, {bathrooms}-bath apartment in NYC is estimated at ${estimate:,.0f}.'
    explanation = f' In this model, each bedroom adds ${bed_coef:,.0f} & each bathroom adds ${bath_coef:,.0f}.'
    return result + explanation

_**Let's do something similar, but with a tuned Random Forest and Shapley Values.**_


In [ ]:
# Assign to X, y
features = ['bedrooms', 'bathrooms', 'longitude', 'latitude']
target = 'price'
X_train = train[features]
y_train = train[target]
X_test = test[features]
y_test = test[target]

In [ ]:
from scipy.stats import randint, uniform
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import RandomizedSearchCV

param_distributions = { 
    'n_estimators': randint(50, 500), 
    'max_depth': [5, 10, 15, 20, None], 
    'max_features': uniform(0, 1), 
}

search = RandomizedSearchCV(
    RandomForestRegressor(random_state=42), 
    param_distributions=param_distributions, 
    n_iter=5, 
    cv=2, 
    scoring='neg_mean_absolute_error', 
    verbose=10, 
    return_train_score=True, 
    n_jobs=-1, 
    random_state=42
)

search.fit(X_train, y_train);

In [ ]:
print('Best hyperparameters', search.best_params_)
print('Cross-validation MAE', -search.best_score_)
model = search.best_estimator_

## Explain individual predictions with shapley value plots

#### [Dan Becker explains Shapley Values:](https://www.kaggle.com/dansbecker/shap-values)

>You've seen (and used) techniques to extract general insights from a machine learning model. But what if you want to break down how the model works for an individual prediction?

>SHAP Values (an acronym from SHapley Additive exPlanations) break down a prediction to show the impact of each feature. 

>There is some complexity to the technique ... We won't go into that detail here, since it isn't critical for using the technique. [This blog post](https://towardsdatascience.com/one-feature-attribution-method-to-supposedly-rule-them-all-shapley-values-f3e04534983d) has a longer theoretical explanation.


In [ ]:
# Get an individual observation to explain.
# For example, the 0th row from the test set.

row = X_test.iloc[[0]]  # Dataframe with a single row (double brackets keeps it a dataframe)

In [ ]:
# What was the actual rent for this apartment?
y_test.iloc[[0]]

In [ ]:
# What does the model predict for this apartment?
model.predict(row)

In [ ]:
# Why did the model predict this?
# Look at a Shapley Values Force Plot

import shap

explainer = shap.TreeExplainer(model)
shap_values = explainer.shap_values(row)

shap.initjs()
shap.force_plot(
    base_value=explainer.expected_value, 
    shap_values=shap_values,
    features=row
)

In [ ]:
# Base value is approximately equal to the mean baseline
explainer.expected_value, y_train.mean()

In [ ]:
feature_names = row.columns
feature_values = row.values[0]
shaps = pd.Series(shap_values[0], zip(feature_names, feature_values))
shaps

In [ ]:
shaps.sum()

In [ ]:
explainer.expected_value + shaps.sum()

### Define the predict function

In [ ]:
def predict(bedrooms, bathrooms, longitude, latitude):

    # Make dataframe from the inputs
    df = pd.DataFrame(
        data=[[bedrooms, bathrooms, longitude, latitude]], 
        columns=['bedrooms', 'bathrooms', 'longitude', 'latitude']
    )

    # Get the model's prediction
    pred = model.predict(df)[0]

    # Calculate shap values
    explainer = shap.TreeExplainer(model)
    shap_values = explainer.shap_values(df)

    # Get series with shap values, feature names, & feature values
    feature_names = df.columns
    feature_values = df.values[0]
    shaps = pd.Series(shap_values[0], zip(feature_names, feature_values))

    # Print results
    result = f'${pred:,.0f} estimated rent for this NYC apartment. \n\n'
    result += f'Starting from baseline of ${explainer.expected_value:,.0f} \n'
    result += shaps.to_string()
    print(result)


    # Show shapley values force plot
    shap.initjs()
    return shap.force_plot(
        base_value=explainer.expected_value, 
        shap_values=shap_values, 
        features=df
    )

predict(3, 1.5, -73.9425, 40.7145)

In [ ]:
# What if it was a 2 bedroom?
predict(2, 1.5, -73.9425, 40.7145)

In [ ]:
# What if it was a 1 bedroom?
predict(1, 1.5, -73.9425, 40.7145)

## Classification: Lending Club 🏦

This notebook uses Lending Club data, historical and current. Predict if peer-to-peer loans are charged off or fully paid. Decide which loans to invest in.


In [ ]:
import pandas as pd

# Stratified sample, 10% of expired Lending Club loans, grades A-D
# Source: https://www.lendingclub.com/info/download-data.action
history = pd.read_csv(DATA_PATH+'lending-club/lending-club-subset.csv')
history['issue_d'] = pd.to_datetime(history['issue_d'], infer_datetime_format=True)

# Current loans available for manual investing, June 17, 2019
# Source: https://www.lendingclub.com/browse/browse.action
current = pd.read_csv(DATA_PATH+'../data/lending-club/primaryMarketNotes_browseNotes_1-RETAIL.csv')

In [ ]:
# Transform earliest_cr_line to an integer:
# How many days the earliest credit line was open, before the loan was issued.
# For current loans available for manual investing, assume the loan will be issued today.
history['earliest_cr_line'] = pd.to_datetime(history['earliest_cr_line'], infer_datetime_format=True)
history['earliest_cr_line'] = history['issue_d'] - history['earliest_cr_line']
history['earliest_cr_line'] = history['earliest_cr_line'].dt.days

current['earliest_cr_line'] = pd.to_datetime(current['earliest_cr_line'], infer_datetime_format=True)
current['earliest_cr_line'] = pd.Timestamp.today() - current['earliest_cr_line']
current['earliest_cr_line'] = current['earliest_cr_line'].dt.days

# Transform earliest_cr_line for the secondary applicant
history['sec_app_earliest_cr_line'] = pd.to_datetime(history['sec_app_earliest_cr_line'], infer_datetime_format=True, errors='coerce')
history['sec_app_earliest_cr_line'] = history['issue_d'] - history['sec_app_earliest_cr_line']
history['sec_app_earliest_cr_line'] = history['sec_app_earliest_cr_line'].dt.days

current['sec_app_earliest_cr_line'] = pd.to_datetime(current['sec_app_earliest_cr_line'], infer_datetime_format=True, errors='coerce')
current['sec_app_earliest_cr_line'] = pd.Timestamp.today() - current['sec_app_earliest_cr_line']
current['sec_app_earliest_cr_line'] = current['sec_app_earliest_cr_line'].dt.days

# Engineer features for issue date year & month
history['issue_d_year'] = history['issue_d'].dt.year
history['issue_d_month'] = history['issue_d'].dt.month

current['issue_d_year'] = pd.Timestamp.today().year
current['issue_d_month'] = pd.Timestamp.today().month

In [ ]:
# Calculate percent of each loan repaid
history['percent_paid'] = history['total_pymnt'] / history['funded_amnt']

In [ ]:
# Train on the historical data.
# For the target, use `loan_status` ('Fully Paid' or 'Charged Off')
target = 'loan_status'
X = history.drop(columns=target)
y = history[target]

In [ ]:
# Do train/validate/test 3-way split
from sklearn.model_selection import train_test_split

X_trainval, X_test, y_trainval, y_test = train_test_split(
    X, y, test_size=20000, stratify=y, random_state=42)

X_train, X_val, y_train, y_val = train_test_split(
    X_trainval, y_trainval, test_size=20000, 
    stratify=y_trainval, random_state=42)

print('X_train shape', X_train.shape)
print('y_train shape', y_train.shape)
print('X_val shape', X_val.shape)
print('y_val shape', y_val.shape)
print('X_test shape', X_test.shape)
print('y_test shape', y_test.shape)

In [ ]:
# Save the ids for later, so we can look up actual results,
# to compare with predicted results
train_id = X_train['id']
val_id = X_val['id']
test_id = X_test['id']

In [ ]:
# Use Python sets to compare the historical columns & current columns
common_columns = set(history.columns) & set(current.columns)
just_history = set(history.columns) - set(current.columns)
just_current = set(current.columns) - set(history.columns)

In [ ]:
# For features, use only the common columns shared by the historical & current data.
features = list(common_columns)
X_train = X_train[features]
X_val = X_val[features]
X_test = X_test[features]

In [ ]:
def wrangle(X):
    X = X.copy()

    # Engineer new feature for every feature: is the feature null?
    for col in X:
        X[col+'_NULL'] = X[col].isnull()
    
    # Convert percentages from strings to floats
    X['int_rate'] = X['int_rate'].str.strip('%').astype(float)
    X['revol_util'] = X['revol_util'].str.strip('%').astype(float)
    
    # Convert employment length from string to float
    X['emp_length'] = X['emp_length'].str.replace(r'\D','').astype(float)
        
    # Create features for three employee titles: teacher, manager, owner
    X['emp_title'] = X['emp_title'].str.lower()
    X['emp_title_teacher'] = X['emp_title'].str.contains('teacher', na=False)
    X['emp_title_manager'] = X['emp_title'].str.contains('manager', na=False)
    X['emp_title_owner']   = X['emp_title'].str.contains('owner', na=False)

    # Get length of free text fields
    X['title'] = X['title'].str.len()
    X['desc'] = X['desc'].str.len()
    X['emp_title'] = X['emp_title'].str.len()
    
    # Convert sub_grade from string "A1"-"D5" to numbers
    sub_grade_ranks = {'A1': 1.1, 'A2': 1.2, 'A3': 1.3, 'A4': 1.4, 'A5': 1.5, 
                       'B1': 2.1, 'B2': 2.2, 'B3': 2.3, 'B4': 2.4, 'B5': 2.5, 
                       'C1': 3.1, 'C2': 3.2, 'C3': 3.3, 'C4': 3.4, 'C5': 3.5, 
                       'D1': 4.1, 'D2': 4.2, 'D3': 4.3, 'D4': 4.4, 'D5': 4.5}
    X['sub_grade'] = X['sub_grade'].map(sub_grade_ranks)
    
    # Drop some columns
    X = X.drop(columns='id')        # Always unique
    X = X.drop(columns='url')       # Always unique
    X = X.drop(columns='member_id') # Always null
    X = X.drop(columns='grade')     # Duplicative of sub_grade
    X = X.drop(columns='zip_code')  # High cardinality
    
    # Only use these features which had nonzero permutation importances in earlier models    
    features = ['acc_open_past_24mths', 'addr_state', 'all_util', 'annual_inc', 
                'annual_inc_joint', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util', 
                'collections_12_mths_ex_med', 'delinq_amnt', 'desc_NULL', 'dti', 
                'dti_joint', 'earliest_cr_line', 'emp_length', 'emp_length_NULL', 
                'emp_title', 'emp_title_NULL', 'emp_title_owner', 'fico_range_high', 
                'funded_amnt', 'home_ownership', 'inq_last_12m', 'inq_last_6mths', 
                'installment', 'int_rate', 'issue_d_month', 'issue_d_year', 'loan_amnt', 
                'max_bal_bc', 'mo_sin_old_il_acct', 'mo_sin_old_rev_tl_op', 
                'mo_sin_rcnt_rev_tl_op', 'mort_acc', 'mths_since_last_major_derog_NULL', 
                'mths_since_last_record', 'mths_since_recent_bc', 'mths_since_recent_inq', 
                'num_actv_bc_tl', 'num_actv_rev_tl', 'num_op_rev_tl', 'num_rev_tl_bal_gt_0', 
                'num_tl_120dpd_2m_NULL', 'open_rv_12m_NULL', 'open_rv_24m', 
                'pct_tl_nvr_dlq', 'percent_bc_gt_75', 'pub_rec_bankruptcies', 'purpose', 
                'revol_bal', 'revol_bal_joint', 'sec_app_earliest_cr_line', 
                'sec_app_fico_range_high', 'sec_app_open_acc', 'sec_app_open_act_il', 
                'sub_grade', 'term', 'title', 'title_NULL', 'tot_coll_amt', 
                'tot_hi_cred_lim', 'total_acc', 'total_bal_il', 'total_bc_limit', 
                'total_cu_tl', 'total_rev_hi_lim']    
    X = X[features]
    
    # Reset index
    X = X.reset_index(drop=True)
    
    # Return the wrangled dataframe
    return X


X_train = wrangle(X_train)
X_val   = wrangle(X_val)
X_test  = wrangle(X_test)

print('X_train shape', X_train.shape)
print('X_val shape', X_val.shape)
print('X_test shape', X_test.shape)

In [ ]:
import category_encoders as ce
from sklearn.impute import SimpleImputer
from sklearn.pipeline import make_pipeline
from xgboost import XGBClassifier

processor = make_pipeline(
    ce.OrdinalEncoder(), 
    SimpleImputer(strategy='median')
)

X_train_processed = processor.fit_transform(X_train)
X_val_processed = processor.transform(X_val)

eval_set = [(X_train_processed, y_train), 
            (X_val_processed, y_val)]

model = XGBClassifier(n_estimators=1000, n_jobs=-1)
model.fit(X_train_processed, y_train, eval_set=eval_set, eval_metric='auc', 
          early_stopping_rounds=10)

In [ ]:
# THIS CELL ISN'T ABOUT THE NEW OBJECTIVES FOR TODAY
# BUT IT IS IMPORTANT FOR YOUR SPRINT CHALLENGE

from sklearn.metrics import roc_auc_score
X_test_processed = processor.transform(X_test)
class_index = 1
y_pred_proba = model.predict_proba(X_test_processed)[:, class_index]
print(f'Test ROC AUC for class {class_index}:')
print(roc_auc_score(y_test, y_pred_proba)) # Ranges from 0-1, higher is better

### Look at predictions vs actuals


In [ ]:
df = pd.DataFrame({
    'id': test_id, 
    'pred_proba': y_pred_proba, 
    'status_group': y_test
})

df = df.merge(
     history[['id', 'issue_d', 'sub_grade', 'percent_paid', 'term', 'int_rate']], 
     how='left'
)

In [ ]:
df.head()

In [ ]:
fully_paid = df['status_group'] == 'Fully Paid'
charged_off = ~fully_paid
right = (fully_paid) == (df['pred_proba'] > 0.50)
wrong = ~right

#### Loan was fully paid, model's prediction was right


In [ ]:
df[fully_paid & right].sample(n=10, random_state=1).sort_values(by='pred_proba')

In [ ]:
# To explain the prediction for test observation with index #3094, 
# first, get all of the features for that observation
row = X_test.iloc[[3094]]
row

## Explain individual predictions with shapley value plots

In [ ]:
# STUDY/PRACTICE THIS CELL FOR THE SPRINT CHALLENGE
import shap

explainer = shap.TreeExplainer(model)
row_processed = processor.transform(row)
shap_values = explainer.shap_values(row_processed)

shap.initjs()
shap.force_plot(
    base_value=explainer.expected_value, 
    shap_values=shap_values, 
    features=row, 
    link='logit' # For classification, this shows predicted probabilities
)

## BONUS: Make a function to explain predictions

Goal Output:

```
The model predicts this loan is Fully Paid, with 74% probability.
 
 
Top 3 reasons for prediction:
1. dti is 10.97.
2. term is  36 months.
3. total_acc is 45.0.
 
 
Top counter-argument against prediction:
- sub_grade is 4.2.
 
<INSERT SHAPLEY VALUE FORCE PLOT HERE>
```

In [ ]:
feature_names = row.columns
feature_values = row.values[0]
shaps = pd.Series(shap_values[0], zip(feature_names, feature_values))
shaps

In [ ]:
pros = shaps.sort_values(ascending=False)[:3].index
cons = shaps.sort_values(ascending=True)[:3].index

In [ ]:
pros

In [ ]:
cons

In [ ]:
print('Top 3 reasons for fully paid:')
for i, pro in enumerate(pros, start=1):
    feature_name, feature_value = pro
    print(f'{i}. {feature_name} is {feature_value}.')

print('\n')
print('Cons:')
for i, con in enumerate(cons, start=1):
    feature_name, feature_value = con
    print(f'{i}. {feature_name} is {feature_value}.')

In [ ]:
def explain(row_number):
    positive_class = 'Fully Paid'
    positive_class_index = 1

    # Get & process the data for the row
    row = X_test.iloc[[row_number]]
    row_processed = processor.transform(row)

    # Make predictions (includes predicted probability)
    pred = model.predict(row_processed)[0]
    pred_proba = model.predict_proba(row_processed)[0, positive_class_index]
    pred_proba *= 100
    if pred != positive_class:
        pred_proba = 100 - pred_proba

    # Show prediction & probability
    print(f'The model predicts this loan is {pred}, with {pred_proba:.0f}% probability.')
    
    # Get shapley additive explanations
    shap_values = explainer.shap_values(row_processed)

    # Get top 3 "pros & cons" for fully paid
    feature_names = row.columns
    feature_values = row.values[0]
    shaps = pd.Series(shap_values[0], zip(feature_names, feature_values))
    pros = shaps.sort_values(ascending=False)[:3].index
    cons = shaps.sort_values(ascending=True)[:3].index

    # Show top 3 reason for prediction
    print('\n')
    print('Top 3 reasons for prediction:')
    evidence = pros if pred == positive_class else cons
    for i, info in enumerate(evidence, start=1):
        feature_name, feature_value = info
        print(f'{i}. {feature_name} is {feature_value}.')

    # Show top 1 counter-argument against prediction
    print('\n')
    print('Top counter-argument against prediction:')
    evidence = cons if pred == positive_class else pros
    feature_name, feature_value = evidence[0]
    print(f'- {feature_name} is {feature_value}.')

    # Show Shapley Values Force Plot
    shap.initjs()
    return shap.force_plot(
        base_value=explainer.expected_value, 
        shap_values=shap_values, 
        features=row, 
        link='logit' # For classification, this shows predicted probabilities
    )

explain(3094)

## Look at more examples


#### Loan was charged off, model's prediction was right


In [ ]:
df[charged_off & right].sample(n=10, random_state=1).sort_values(by='pred_proba')

In [ ]:
explain(8383)

#### Loan was fully paid, model's prediction was wrong


In [ ]:
df[fully_paid & wrong].sample(n=10, random_state=1).sort_values(by='pred_proba')

In [ ]:
explain(18061)

In [ ]:
explain(6763)

#### Loan was charged off, model's prediction was wrong


In [ ]:
df[charged_off & wrong].sample(n=10, random_state=1).sort_values(by='pred_proba')

In [ ]:
explain(19883)

# Checklist

Don't let the perfect be the enemy of the good.

## The Good

- Publish a blog post

- Choose a 
  - target, (**Poisonous or edible**)
  - evaluation metric (**I shoose recall**), and 
  - baseline (**53.3% chance of being correct if we always guess 'edible'**)
    
  for a 
  - regression, 
  - binary classification **(THIS IS WHAT I CHOOSE)**, or 
  - multi-class classification problem, 
  
  with their labeled, tabular dataset.
  
- Fit and evaluate
  - any linear model 
  
  for 
  
  - regression or 
  - classification **(THIS IS WHAT I CHOOSE)**
  
- Fit and evaluate 
  - a decision tree, 
  - random forest, or 
  - gradient boosting model 
  
  for regression or classification
  
  Student writes 300+ words (not including code). 
  
  Student reports baseline score (**baseline is 53.3%**)
  validation scores from 2+ models
  - model 1
  - model 2
  and test score from 1 selected model
  
  
  Student makes 2+ visualizations to explain their model
  - viz one
  
  Student commits all code to GitHub. Notebooks have no error messages
  
  

## The Perfect